lms/djangoapps/courseware/tabs.py

```diff
@@ -3,44 +3,347 @@
 """
 from django.conf import settings
-from django.test.client import RequestFactory
 from django.utils.translation import ugettext as _
 
 from courseware.access import has_access
 from courseware.entrance_exams import user_must_complete_entrance_exam
-from student.models import CourseEnrollment, EntranceExamConfiguration
-from xmodule.tabs import CourseTabList
+from openedx.core.djangoapps.course_views.course_views import CourseViewTypeManager, CourseViewType, StaticTab
+from student.models import CourseEnrollment
+from xmodule.tabs import CourseTab, CourseTabList, key_checker
 
-from util import milestones_helpers
+
+class EnrolledCourseViewType(CourseViewType):
+    """
+    A base class for any view types that require a user to be enrolled.
+    """
+    @classmethod
+    def is_enabled(cls, course, user=None):
+        if not user:
+            return True
+        return CourseEnrollment.is_enrolled(user, course.id) or has_access(user, 'staff', course, course.id)
+
+
+class CoursewareViewType(EnrolledCourseViewType):
+    """
+    The main courseware view.
+    """
+    name = 'courseware'
+    title = _('Courseware')
+    priority = 10
+    view_name = 'courseware'
+    is_movable = False
+
+
+class CourseInfoViewType(CourseViewType):
+    """
+    The course info view.
+    """
+    name = 'course_info'
+    title = _('Course Info')
+    priority = 20
+    view_name = 'info'
+    tab_id = 'info'
+    is_movable = False
+
+    @classmethod
+    def is_enabled(cls, course, user=None):
+        return True
 
 
-def get_course_tab_list(course, user):
+class SyllabusCourseViewType(EnrolledCourseViewType):
+    """
+    A tab for the course syllabus.
+    """
+    name = 'syllabus'
+    title = _('Syllabus')
+    priority = 30
+    view_name = 'syllabus'
+
+    @classmethod
+    def is_enabled(cls, course, user=None):  # pylint: disable=unused-argument
+        if not super(SyllabusCourseViewType, cls).is_enabled(course, user=user):
+            return False
+        return hasattr(course, 'syllabus_present') and course.syllabus_present
+
+
+class ProgressCourseViewType(EnrolledCourseViewType):
+    """
+    The course progress view.
+    """
+    name = 'progress'
+    title = _('Progress')
+    priority = 40
+    view_name = 'progress'
+    is_hideable = True
+
+    @classmethod
+    def is_enabled(cls, course, user=None):  # pylint: disable=unused-argument
+        if not super(ProgressCourseViewType, cls).is_enabled(course, user=user):
+            return False
+        return not course.hide_progress_tab
+
+
+class TextbookCourseViewsBase(CourseViewType):
+    """
+    Abstract class for textbook collection tabs classes.
+    """
+    # Translators: 'Textbooks' refers to the tab in the course that leads to the course' textbooks
+    title = _("Textbooks")
+    is_collection = True
+
+    @classmethod
+    def is_enabled(cls, course, user=None):  # pylint: disable=unused-argument
+        return user is None or user.is_authenticated()
```

Comment:
So, it's visible if there's no user object, or if the user is logged in, but not if the user is an `AnonymousUser`? That seems odd.